File: ChatClientCS/Commands/RelayCommandAsync.cs
using System;
using System.Threading.Tasks;
using System.Windows.Input;

namespace ChatClientCS.Commands
{
    public class RelayCommandAsync : ICommand
    {
        private readonly Func<Task> _esegui;
        private readonly Predicate<object> _puoEseguire;
        private bool staEseguendo;

        public RelayCommandAsync(Func<Task> esegui) : this(esegui, null) { }

        public RelayCommandAsync(Func<Task> esegui, Predicate<object> canExecute)
        {
            _esegui = esegui;
            _puoEseguire = canExecute;
        }

        public bool CanExecute(object parametro)
        {
            if (!staEseguendo && _puoEseguire == null) return true;
            return (!staEseguendo && _puoEseguire(parametro));
        }

        public event EventHandler CanExecuteChanged
        {
            add { CommandManager.RequerySuggested += value; }
            remove { CommandManager.RequerySuggested -= value; }
        }

        public async void Execute(object parameter)
        {
            staEseguendo = true;
            try { await _esegui(); }
            finally { staEseguendo = false; }
        }
    }
}

File: ChatClientVB/Commands/RelayCommandAsync.vb
Public Class RelayCommandAsync
    Implements ICommand

    Private ReadOnly _esegui As Func(Of Task)
    Private ReadOnly _puoEseguire As Predicate(Of Object)
    Private inEsecuzione As Boolean

    Public Sub New(esegui As Func(Of Task))
        Me.New(esegui, Nothing)
    End Sub

    Public Sub New(esegui As Func(Of Task), eseguibile As Predicate(Of Object))
        _esegui = esegui
        _puoEseguire = eseguibile
    End Sub

    Public Function CanExecute(parametro As Object) As Boolean Implements ICommand.CanExecute
        If Not inEsecuzione AndAlso _puoEseguire Is Nothing Then Return True
        Return Not inEsecuzione AndAlso _puoEseguire(parametro)
    End Function

    Public Custom Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged
        AddHandler(value As EventHandler)
            AddHandler CommandManager.RequerySuggested, value
        End AddHandler

        RemoveHandler(value As EventHandler)
            RemoveHandler CommandManager.RequerySuggested, value
        End RemoveHandler

        RaiseEvent(sender As Object, e As EventArgs)
        End RaiseEvent
    End Event

    Public Async Sub Execute(parametro As Object) Implements ICommand.Execute
        inEsecuzione = True
        Try
            Await _esegui()
        Finally
            inEsecuzione = False
        End Try
    End Sub
End Class
